File: src/logic/selector.ts
/**
 * Core spectrometer selector logic.
 *
 * Given a wavelength range and target resolution, finds all Evolve
 * spectrometer configurations that fully cover the range, recommending
 * the largest slit width that still meets the resolution spec
 * (maximising optical throughput).
 *
 * If no exact matches exist, returns "near misses" — configs that cover
 * the wavelength range but cannot achieve the resolution even at the
 * narrowest slit.
 */
import type {
  ResolutionRecord,
  GratingOverrides,
  CodeInfo,
  CodeRangeLookup,
} from "../types/spectrometer";

/** Grating codes grouped by blaze wavelength (nm). */
export type BlazeCodeMap = Record<number, CodeInfo[]>;

/** Enriched result record carried through Card / Compare. */
export interface EnrichedResult {
  /** Original resolution record fields */
  platforms: string[];
  evolveNames: string[];
  isEvolve: boolean;
  gratingGrooves: number;
  blazeWavelengths: number[];
  bandwidthNm: number;
  selectableRange: [number, number];
  model?: string;

  /** Computed fields */
  recSlit: number;
  recRes: number;
  codesByBlaze: BlazeCodeMap;
  allSlits: [number, number][]; // sorted ascending by slit width
}

export interface SelectorSearchResult {
  matches: EnrichedResult[];
  nearMisses: EnrichedResult[];
}

/**
 * Look up grating codes grouped by blaze wavelength, enriched with
 * each code's configured wavelength window from the naming table.
 * Includes ALL blaze wavelengths so the UI can annotate out-of-range ones.
 */
export function lookupGratingCodesByBlaze(
  record: ResolutionRecord,
  overrides: GratingOverrides,
  codeRangeLookup: CodeRangeLookup,
): BlazeCodeMap {
  const result: BlazeCodeMap = {};
  for (const blaze of record.blazeWavelengths) {
    const codeStrs: string[] = [];
    for (const platform of record.platforms) {
      const key = `${platform}|${record.gratingGrooves}|${blaze}`;
      const found = overrides[key];
      if (found) codeStrs.push(...found);
    }
    const unique = [...new Set(codeStrs)];
    if (unique.length > 0) {
      result[blaze] = unique.map((code) => {
        const range = codeRangeLookup[code];
        return {
          code,
          wlMin: range ? range[0] : null,
          wlMax: range ? range[1] : null,
        };
      });
    }
  }
  return result;
}

/**
 * Format a part number: XXxxxx-SSS-GGGG
 *   XX   = OtO platform code (e.g. "SE", "SW")
 *   xxxx = model placeholder (customer specifies actual model)
 *   SSS  = slit width in µm, zero-padded to 3 digits
 *   GGGG = grating code (e.g. "DUV5", "NIRC")
 */
export function formatPartNumber(platform: string, slitUm: number, gratingCode: string): string {
  return `${platform}xxxx-${String(slitUm).padStart(3, "0")}-${gratingCode}`;
}

/** Returns true if a blaze wavelength falls inside the user's search window. */
export function blazeInRange(blaze: number, wlMin: number, wlMax: number): boolean {
  return blaze >= wlMin && blaze <= wlMax;
}

/**
 * Sort comparator that orders blaze wavelengths (as numeric strings) with
 * in-range blazes first, then ascending by wavelength. Used by ResultCard
 * and CompareTable to present grating codes in a consistent order.
 */
export function blazeSortComparator(wlMin: number, wlMax: number) {
  return (a: string, b: string): number => {
    const aIn = blazeInRange(Number(a), wlMin, wlMax);
    const bIn = blazeInRange(Number(b), wlMin, wlMax);
    if (aIn !== bIn) return aIn ? -1 : 1;
    return Number(a) - Number(b);
  };
}

/** Count blaze wavelengths in this result that fall within the search range. */
function countInRangeBlazes(r: { codesByBlaze: BlazeCodeMap }, wlMin: number, wlMax: number): number {
  return Object.keys(r.codesByBlaze)
    .map(Number)
    .filter(b => blazeInRange(b, wlMin, wlMax)).length;
}

/**
 * Generate a stable unique ID for a record (used for checkbox state).
 */
export function recordId(r: EnrichedResult): string {
  return `${r.platforms.join("")}-${r.gratingGrooves}-${r.blazeWavelengths.join("")}-${r.model ?? ""}`;
}

/**
 * Main search function. Finds all spectrometer configurations that fully
 * cover the requested wavelength range, recommends the largest slit width
 * meeting the resolution spec, and returns results sorted by throughput
 * with in-range blaze wavelengths prioritised.
 *
 * @param wlMin           - minimum wavelength required (nm)
 * @param wlMax           - maximum wavelength required (nm)
 * @param maxRes          - maximum acceptable resolution (nm, smaller = better)
 * @param records         - resolution records to search
 * @param overrides       - grating code override table
 * @param codeRangeLookup - grating code → wavelength window lookup
 */
export function search(
  wlMin: number,
  wlMax: number,
  maxRes: number,
  records: ResolutionRecord[],
  overrides: GratingOverrides,
  codeRangeLookup: CodeRangeLookup = {},
): SelectorSearchResult {
  const requiredBandwidth = wlMax - wlMin;
  const matches: EnrichedResult[] = [];
  const nearMisses: EnrichedResult[] = [];

  for (const r of records) {
    // Filter: must have selectable range and bandwidth
    if (!r.selectableRange || r.bandwidthNm === null) continue;

    // Filter: bandwidth must accommodate the requested range
    if (r.bandwidthNm < requiredBandwidth) continue;

    // Filter: selectable range must fully cover both endpoints
    if (r.selectableRange[0] > wlMin || r.selectableRange[1] < wlMax) continue;

    // Build sorted slit array: [[slitUm, resolutionNm], ...]
    const slitEntries: [number, number][] = Object.entries(r.slitResolutions)
      .map(([k, v]) => [Number(k), v] as [number, number])
      .sort((a, b) => a[0] - b[0]); // ascending by slit width

    if (slitEntries.length === 0) continue;

    // Look up grating codes grouped by blaze wavelength (all blazes)
    const codesByBlaze = lookupGratingCodesByBlaze(r, overrides, codeRangeLookup);

    // Build the enriched base record
    const base = {
      platforms: r.platforms,
      evolveNames: r.evolveNames,
      isEvolve: r.isEvolve,
      gratingGrooves: r.gratingGrooves,
      blazeWavelengths: r.blazeWavelengths,
      bandwidthNm: r.bandwidthNm,
      selectableRange: r.selectableRange,
      model: r.model,
      codesByBlaze,
      allSlits: slitEntries,
    };

    // Find the largest slit that achieves the required resolution
    // (search from largest to smallest)
    let bestSlit: [number, number] | null = null;
    const descSlits = [...slitEntries].sort((a, b) => b[0] - a[0]);
    for (const [slitUm, res] of descSlits) {
      if (res <= maxRes) {
        bestSlit = [slitUm, res];
        break;
      }
    }

    if (bestSlit) {
      matches.push({ ...base, recSlit: bestSlit[0], recRes: bestSlit[1] });
    } else {
      // Near miss: use the narrowest slit (best resolution achievable)
      const narrowest = slitEntries[0]; // smallest slit = best resolution
      nearMisses.push({ ...base, recSlit: narrowest[0], recRes: narrowest[1] });
    }
  }

  // Sort matches: largest slit (throughput) first, then in-range blaze count, then resolution.
  // Throughput is the primary concern — a 60µm slit with 0.9nm resolution is always
  // preferable to a 10µm slit with 1.0nm resolution, regardless of blaze alignment.
  matches.sort((a, b) => {
    // Primary: largest recommended slit (best throughput)
    if (a.recSlit !== b.recSlit) return b.recSlit - a.recSlit;
    // Secondary: more in-range blazes is better
    const aIn = countInRangeBlazes(a, wlMin, wlMax);
    const bIn = countInRangeBlazes(b, wlMin, wlMax);
    if (aIn !== bIn) return bIn - aIn;
    // Tertiary: better resolution as tiebreaker
    return a.recRes - b.recRes;
  });

  // Sort near misses: in-range blaze first, then closest achievable resolution
  nearMisses.sort((a, b) => {
    const aIn = countInRangeBlazes(a, wlMin, wlMax);
    const bIn = countInRangeBlazes(b, wlMin, wlMax);
    if ((aIn > 0) !== (bIn > 0)) return bIn > 0 ? 1 : -1;
    return a.recRes - b.recRes;
  });

  return { matches, nearMisses };
}

File: data/selector.py
import json

with open('/home/claude/spectrometer_data.json') as f:
    data = json.load(f)

resolution_records = data['resolution_records']
naming_records = data['naming_records']
evolve_map = data['evolve_map']


def find_spectrometers(wl_min, wl_max, max_resolution, evolve_only=True):
    """
    Find spectrometer configurations that:
    1. Fully cover the requested wavelength range
    2. Achieve the requested resolution or better
    3. Use the largest slit possible (best throughput)
    
    Args:
        wl_min: minimum wavelength needed (nm)
        wl_max: maximum wavelength needed (nm)
        max_resolution: maximum acceptable resolution (nm) - smaller = better
        evolve_only: if True, only return Evolve-branded products
    
    Returns:
        List of matching configurations, sorted by best throughput (largest slit)
    """
    required_bandwidth = wl_max - wl_min
    matches = []
    
    for rec in resolution_records:
        # Filter: Evolve only
        if evolve_only and not rec.get('is_evolve', False):
            continue
        
        # Filter: selectable range must fully cover requested range
        sel = rec.get('selectable_range')
        if not sel:
            continue
        sel_min, sel_max = sel
        
        # The selectable range is the full range the grating+bench CAN cover.
        # The actual configured range must fit within this.
        # Check: can the grating's bandwidth accommodate the requested range
        # AND does the selectable range cover both endpoints?
        bandwidth = rec.get('bandwidth_nm')
        if bandwidth is None:
            continue
        
        # The requested range must fit within the bandwidth
        if required_bandwidth > bandwidth:
            continue
        
        # The selectable range must cover the requested endpoints
        if sel_min > wl_min or sel_max < wl_max:
            continue
        
        # Now find the largest slit that achieves the required resolution
        slit_res = rec.get('slit_resolutions', {})
        if not slit_res:
            continue
        
        # Sort slits from largest to smallest
        sorted_slits = sorted(slit_res.items(), key=lambda x: -int(x[0]))
        
        best_slit = None
        best_res = None
        for slit_str, res in sorted_slits:
            slit = int(slit_str)
            if res <= max_resolution:
                best_slit = slit
                best_res = res
                break
        
        if best_slit is None:
            # No slit meets the resolution requirement
            # But include the best (smallest slit) for info if it's close
            continue
        
        # Look up grating code from naming rules
        grating_codes = find_grating_codes(rec)
        
        matches.append({
            'platforms': rec['platforms'],
            'evolve_names': rec.get('evolve_names', []),
            'grating_grooves': rec['grating_grooves'],
            'blaze_wavelengths': rec['blaze_wavelengths'],
            'bandwidth_nm': bandwidth,
            'selectable_range': sel,
            'recommended_slit_um': best_slit,
            'resolution_at_slit_nm': best_res,
            'all_slit_resolutions': slit_res,
            'grating_codes': grating_codes,
            'model': rec.get('model'),
        })
    
    # Sort by: largest slit first (best throughput), then best resolution
    matches.sort(key=lambda x: (-x['recommended_slit_um'], x['resolution_at_slit_nm']))
    
    return matches


def find_grating_codes(rec):
    """Cross-reference a resolution record to the naming rule to find grating codes."""
    codes = []
    grooves = rec['grating_grooves']
    platforms = set(rec['platforms'])
    
    for nr in naming_records:
        if nr['grooves'] is None:
            continue
        # Match by groove density (allow small float tolerance)
        if abs(nr['grooves'] - grooves) > 1:
            continue
        # Check if any platform matches
        nr_platforms = set(nr['platform_codes'])
        if not platforms.intersection(nr_platforms):
            # Also check raw string for partial matches
            raw = nr['platforms_raw']
            found = False
            for p in platforms:
                if p in raw:
                    found = True
                    break
            if not found:
                continue
        
        # Check if blaze wavelength matches (if available)
        if nr['blaze'] is not None and rec['blaze_wavelengths']:
            if nr['blaze'] not in rec['blaze_wavelengths']:
                continue
        
        codes.append({
            'code': nr['code'],
            'wl_range': f"{nr['wl_min']}-{nr['wl_max']}nm" if nr['wl_min'] else nr['wl_range_str'],
            'grating_spec': nr['grating_spec_str'],
            'platforms_raw': nr['platforms_raw'],
        })
    
    return codes


def format_results(matches, wl_min, wl_max, max_resolution):
    """Pretty print results."""
    print(f"\n{'='*80}")
    print(f"SPECTROMETER SELECTOR RESULTS")
    print(f"Requested: {wl_min}-{wl_max} nm, resolution ≤ {max_resolution} nm")
    print(f"Required bandwidth: {wl_max - wl_min} nm")
    print(f"{'='*80}")
    
    if not matches:
        print("\n  No matching configurations found.")
        print("  Consider relaxing resolution requirement or narrowing wavelength range.")
        return
    
    print(f"\n  Found {len(matches)} matching configuration(s):\n")
    
    for i, m in enumerate(matches, 1):
        evolve_str = ' / '.join(m['evolve_names']) if m['evolve_names'] else ', '.join(m['platforms'])
        model_str = f" ({m['model']})" if m.get('model') else ''
        
        print(f"  {'─'*70}")
        print(f"  #{i}  {evolve_str}{model_str}")
        print(f"       Grating: {m['grating_grooves']} g/mm, blazed at {m['blaze_wavelengths']} nm")
        print(f"       Bandwidth: {m['bandwidth_nm']} nm | Selectable: {m['selectable_range'][0]}-{m['selectable_range'][1]} nm")
        print(f"       ★ Recommended slit: {m['recommended_slit_um']} µm → {m['resolution_at_slit_nm']} nm resolution")
        
        # Show all slit options
        slit_str = ' | '.join(f"{int(k)}µm: {v}nm" for k, v in 
                              sorted(m['all_slit_resolutions'].items(), key=lambda x: int(x[0])))
        print(f"       All slits: {slit_str}")
        
        if m['grating_codes']:
            codes_str = ', '.join(f"{c['code']} ({c['wl_range']})" for c in m['grating_codes'][:5])
            print(f"       Grating codes: {codes_str}")
        else:
            print(f"       Grating codes: (no exact match found in naming table)")
        print()


# ============================================================
# TEST CASES
# ============================================================
if __name__ == '__main__':
    # Example from Steve: 200-550 nm, resolution ≤ 0.5 nm
    print("\n\n" + "█"*80)
    print("TEST 1: Customer needs 200-550 nm, resolution ≤ 0.5 nm")
    print("█"*80)
    results = find_spectrometers(200, 550, 0.5)
    format_results(results, 200, 550, 0.5)
    
    # Test 2: Broader visible range
    print("\n\n" + "█"*80)
    print("TEST 2: Customer needs 380-780 nm (full visible), resolution ≤ 2 nm")
    print("█"*80)
    results = find_spectrometers(380, 780, 2.0)
    format_results(results, 380, 780, 2.0)
    
    # Test 3: NIR range
    print("\n\n" + "█"*80)
    print("TEST 3: Customer needs 900-1700 nm (NIR), resolution ≤ 10 nm")
    print("█"*80)
    results = find_spectrometers(900, 1700, 10.0)
    format_results(results, 900, 1700, 10.0)
    
    # Test 4: Very high resolution UV
    print("\n\n" + "█"*80)
    print("TEST 4: Customer needs 200-400 nm (UV), resolution ≤ 0.3 nm")
    print("█"*80)
    results = find_spectrometers(200, 400, 0.3)
    format_results(results, 200, 400, 0.3)
